modules/internal/SharedObject.chpl

```diff
@@ -255,10 +255,24 @@ module SharedObject {
     //   var s : shared = ownedThing;
     pragma "no doc"
     proc init=(pragma "nil from arg" in take: owned) {
+      var p = take.release();
+
+      this.chpl_t = if this.type.chpl_t != ?
+                    then this.type.chpl_t
+                    else _to_borrowed(p.type);
+
+      var rc:unmanaged ReferenceCount? = nil;
+
+      if p != nil then
+        rc = new unmanaged ReferenceCount();
+
+      this.chpl_p = p;
+      this.chpl_pn = rc;
+
+      this.complete();
+
```

Comment:
Was the motivation for this to put this logic before `compilerError` and you couldn't call `this.init` before the `compilerError` call?

Reply:
It seemed easier to get reasonable error messages with this code structure, yes. If we called this.init before the compilerError we might have an error within this.init. Anyway, because of the way that `init=` can convey type information through `this.type` it doesn't seem to me that calling `this.init` is good enough.